Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp

```diff
@@ -277,78 +269,118 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
     // explicitly allow the following allocation(s).
     DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
 
-    if (!m_currentAudioSamplingRate) {
-        m_currentAudioSamplingRate = description.sampleRate();
+    if (m_currentStreamDescription != description) {
+        LOG(MediaStream, "description changed");
+        m_currentStreamDescription = toCAAudioStreamDescription(description);
         Locker locker { m_ringBufferLock };
-        m_ringBuffer = InProcessCARingBuffer::allocate(*std::get<const AudioStreamBasicDescription*>(description.platformDescription().description), description.sampleRate() * 2); // Allow for 2s of audio.
+        m_ringBuffers.append(InProcessCARingBuffer::allocate(*std::get<const AudioStreamBasicDescription*>(description.platformDescription().description), description.sampleRate() * 2));
         queueSingleton().dispatch([weakThis = ThreadSafeWeakPtr { *this }, description = *std::get<const AudioStreamBasicDescription*>(description.platformDescription().description)] {
             if (RefPtr protectedThis = weakThis.get())
-                protectedThis->audioSamplesAvailableStarted(description);
+                protectedThis->audioSamplesDescriptionChanged(description);
         });
-        m_originalDescription = toCAAudioStreamDescription(description);
+        m_currentAudioSampleCount = 0;
     }
-    ASSERT(m_originalDescription == description, "sampling rate should never change over a session");
-    if (m_originalDescription != description) {
-        RELEASE_LOG_ERROR(MediaStream, "MediaRecorderPrivateEncoder::appendAudioSampleBuffer sampling rate changed from %u to %f, abort recording", static_cast<uint32_t>(m_currentAudioSamplingRate), description.sampleRate());
-        callOnMainThread([weakThis = ThreadSafeWeakPtr { *this }] {
-            if (RefPtr protectedThis = weakThis.get())
-                protectedThis->stopRecording();
-        });
-        return;
-    }
-    m_lastEnqueuedAudioSampleCount = m_currentAudioSampleCount.load();
-    m_currentAudioSampleCount += sampleCount;
+
+    auto currentAudioTime = m_currentAudioTime;
+    m_lastEnqueuedAudioTimeUs = m_currentAudioTime.toMicroseconds();
+
+    m_currentAudioTime += MediaTime(sampleCount, description.sampleRate());
+    m_currentAudioTimeUs = m_currentAudioTime.toMicroseconds();
+
     {
-        if (!m_ringBufferLock.tryLock())
-            return;
-        Locker locker { AdoptLock, m_ringBufferLock };
-        if (!m_ringBuffer)
+        Locker locker { m_ringBufferLock };
+        if (m_ringBuffers.isEmpty() || !m_ringBuffers.last())
             return;
-        m_ringBuffer->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_lastEnqueuedAudioSampleCount);
+        m_ringBuffers.last()->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_currentAudioSampleCount);
```

Comment:
To make this clearer, I would tend to introduce routines:
```
void addRingBuffer(const AudioStreamDescription&);
void writeDataToRingBuffer(...);
void updateRingBufferIfNeeded();
```
This could also reduce the scope of `m_ringBufferLock ` (which I would rename to `m_ringBuffersLock`) maybe.